pl_bolts/datasets/array_dataset.py

```diff
@@ -0,0 +1,51 @@
+from typing import Tuple
+
+import torch
+from pytorch_lightning.utilities import exceptions
+from torch import Tensor
+from torch.utils.data import Dataset
+
+
+class ArrayDataset(Dataset):
```

Comment:
What if we make an ArrayDataset that can have several features and targets?
Not sure if that's what we are aiming for here but I would love to have the freedom to use something like:
```python
class ArrayDataset(Dataset):
    def __init__(self, *args) -> None:
        self.dataset = apply_to_collection(args, dtype=(np.ndarray, list), function=torch.tensor)

        if not all(tensor.size(0) == self.dataset[0].size(0) for tensor in self.dataset):
            raise exceptions.MisconfigurationException("Shape mismatch between arrays in the first dimension")

    def __len__(self) -> int:
        return self.dataset[0].size(0)

    def __getitem__(self, idx: int) -> list:
        outputs = []
        for tensor in self.dataset:
            tensor = tensor[idx]
            if not ((tensor.dtype == torch.int32) or (tensor.dtype == torch.int64)):
                tensor = tensor.float()
            outputs.append(tensor)

        return outputs
```
Would love to hear your thoughts! @otaj @Ce11an

Reply:
> What if we make an ArrayDataset that can have several features and targets? Not sure if that's what we are aiming for here but I would love to have the freedom to use something like:
> 
> ```python
> class ArrayDataset(Dataset):
>     def __init__(self, *args) -> None:
>         self.dataset = apply_to_collection(args, dtype=(np.ndarray, list), function=torch.tensor)
> 
>         if not all(tensor.size(0) == self.dataset[0].size(0) for tensor in self.dataset):
>             raise exceptions.MisconfigurationException("Shape mismatch between arrays in the first dimension")
> 
>     def __len__(self) -> int:
>         return self.dataset[0].size(0)
> 
>     def __getitem__(self, idx: int) -> list:
>         outputs = []
>         for tensor in self.dataset:
>             outputs.append(tensor[idx])
> 
>         return outputs
> ```
> 
> Would love to hear your thoughts! @otaj @Ce11an

Happy to implement something like this! I did want to make it general. However, as in the original `SklearnDataset` it converted features to floats, I kept the same behaviour. It looks like we are moving away from that completely now so I am all for the change! I haven't used `apply_to_coleection` before. Will check it out 👍🏻

Reply:
Please see the latest commit 👍🏻